contest/contest_info.py

```diff
@@ -0,0 +1,76 @@
+    The above copyright notice and this permission notice shall be included in all
+    copies or substantial portions of the Software.
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+    '''
+from contest.models import Contest
+from contest.models import Contestant
+from contest.models import Clarification
+
+from contest.scoreboard import Scoreboard
+from contest.scoreboard import User
+from contest.scoreboard import ScoreboardProblem
+from contest.scoreboard import UserProblem
+from contest.scoreboard import Submission as ScoreboardSubmission
+
+from problem.models import Problem
+from problem.models import Testcase
+from problem.models import Submission
+from problem.models import SubmissionDetail
+
+def get_contestant_list(contest):
+    return Contestant.objects.filter(contest = contest)
+
+def get_total_testcases(problem):
+    testcases = Testcase.objects.filter(problem = problem)
+    return len(testcases)
+
+def get_contestant_problem_submission_list(contest,contestant,problem):
+    return Submission.objects.filter(problem = problem, submit_time__lte = contest.end_time,
+        submit_time__gte = contest.start_time,user = contestant.user).order_by('submit_time')
+
+def get_passed_testcases(submission):
+    passed_testcases = SubmissionDetail.objects.filter(sid = submission, virdect = SubmissionDetail.AC)
+    return len(passed_testcases)
+
+def get_scoreboard(contest):
+    contestants = get_contestant_list(contest)
+    
+    scoreboard = Scoreboard(contest.start_time)
+    for problem in contest.problem.all():
+        total_testcases = get_total_testcases(problem);
+        new_problem = ScoreboardProblem(problem.id,problem.pname,total_testcases)
+        scoreboard.add_problem(new_problem)
+
+    for contestant in contestants:
+        new_contestant = User(contestant.user.username)
+        for problem in contest.problem.all():
+            submissions = get_contestant_problem_submission_list(contest,contestant,problem)    
+            total_testcases = get_total_testcases(problem)
+            new_problem = UserProblem(problem.id,total_testcases)
+            for submission in submissions:
+                passed_testcases = get_passed_testcases(submission)
+                new_submission = ScoreboardSubmission(submission.submit_time,passed_testcases)
+                new_problem.add_submission(new_submission)
+                if new_submission.is_solved(total_testcases):
+                    break
+            if new_problem.is_solved():
```

Comment:
I suggest to pass 'scoreboard.get_problem(problem.id).total_testcase' as parameter to problem.is_solved, so that you don't have to keep 'total_testcase' in each UserProblem object.
To avoid redundant get_problem callings, you are suggested to exchange order of two for loops(go through problem list first).